src/QtUtils/FutureWatcherTest.cpp

```diff
@@ -0,0 +1,209 @@
+  EXPECT_EQ(watcher.WaitFor(std::move(future), std::chrono::milliseconds{20}),
+            FutureWatcher::Reason::kTimeout);
+}
+
+TEST(FutureWatcher, WaitForWithAbort) {
+  QCoreApplication app{argc, nullptr};
+
+  orbit_base::Promise<void> promise{};
+  orbit_base::Future<void> future = promise.GetFuture();
+
+  QTimer completion_timer{};
+  QObject::connect(&completion_timer, &QTimer::timeout, [&]() { promise.MarkFinished(); });
+  completion_timer.start(std::chrono::milliseconds{40});
+
+  FutureWatcher watcher{};
+
+  QTimer abort_timer{};
+  QObject::connect(&abort_timer, &QTimer::timeout, &watcher, &FutureWatcher::Abort);
+  abort_timer.start(std::chrono::milliseconds{10});
+
+  EXPECT_EQ(watcher.WaitFor(std::move(future), std::chrono::milliseconds{20}),
+            FutureWatcher::Reason::kAbortRequested);
+}
+
+TEST(FutureWatcher, WaitForWithThreadPool) {
+  // One background job which is supposed to succeed. (No timeout, no abort)
+
+  QCoreApplication app{argc, nullptr};
+
+  constexpr size_t kThreadPoolMinSize = 1;
+  constexpr size_t kThreadPoolMaxSize = 2;
+  constexpr absl::Duration kThreadTtl = absl::Milliseconds(5);
+  std::unique_ptr<ThreadPool> thread_pool =
+      ThreadPool::Create(kThreadPoolMinSize, kThreadPoolMaxSize, kThreadTtl);
+
+  absl::Mutex mutex;
+
+  mutex.Lock();
+  orbit_base::Future<void> future = thread_pool->Schedule([&]() { absl::MutexLock lock(&mutex); });
+
+  EXPECT_TRUE(future.IsValid());
+  EXPECT_FALSE(future.IsFinished());
+
+  // The lambda will be executed by the event loop, running inside of watcher.WaitFor.
+  QTimer::singleShot(std::chrono::milliseconds{5}, [&]() { mutex.Unlock(); });
+
+  FutureWatcher watcher{};
+  const auto reason = watcher.WaitFor(std::move(future), std::nullopt);
+
+  EXPECT_EQ(reason, FutureWatcher::Reason::kFutureCompleted);
+
+  thread_pool->ShutdownAndWait();
+}
+
+TEST(FutureWatcher, WaitForWithThreadPoolAndTimeout) {
+  // One background job which is supposed to time out.
+
+  QCoreApplication app{argc, nullptr};
+
+  constexpr size_t kThreadPoolMinSize = 1;
+  constexpr size_t kThreadPoolMaxSize = 2;
+  constexpr absl::Duration kThreadTtl = absl::Milliseconds(5);
+  std::unique_ptr<ThreadPool> thread_pool =
+      ThreadPool::Create(kThreadPoolMinSize, kThreadPoolMaxSize, kThreadTtl);
+
+  absl::Mutex mutex;
+
+  mutex.Lock();
+  orbit_base::Future<void> future = thread_pool->Schedule([&]() { absl::MutexLock lock(&mutex); });
+
+  EXPECT_TRUE(future.IsValid());
+  EXPECT_FALSE(future.IsFinished());
+
+  // The lambda will be executed by the event loop, running inside of watcher.WaitFor.
+  QTimer::singleShot(std::chrono::milliseconds{20}, [&]() { mutex.Unlock(); });
+
+  FutureWatcher watcher{};
+  const auto reason = watcher.WaitFor(std::move(future), std::chrono::milliseconds{5});
+
+  EXPECT_EQ(reason, FutureWatcher::Reason::kTimeout);
+
+  mutex.Unlock();
+  thread_pool->ShutdownAndWait();
+}
+
+TEST(FutureWatcher, WaitForAllWithThreadPool) {
```

Comment:
Maybe a test `WaitForAll({}, std::nullopt)` (empty input) is a good idea.

Reply:
Good idea. Done.